feat: List Section

sooinc opened the pull request:
### Overview
Adds List Section component. This component is built on top of the Page Section component. It takes in a title, an array of elements, a render function and an initial display amount. It has a built in "Show All" functionality.

See Storybook: https://608816dd8ba134699e4ee53f--gamut-preview.netlify.app/storybook/?path=/docs/molecules-listsection--list-section

See used in monolith: https://github.com/codecademy-engineering/Codecademy/pull/23989

<!--- CHANGELOG-DESCRIPTION -->
- Add ListSection molecule
<!--- END-CHANGELOG-DESCRIPTION -->

### PR Checklist

- [x] Related to designs: https://www.figma.com/file/PG0kVYDIuLk3hE566beZUV/Profiles-(Concept-Refinement)?node-id=315%3A2374
- [x] Related to JIRA ticket: Egg-1069
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

<!--
Merging your changes

1. Follow the [PR Title Guide](https://github.com/Codecademy/client-modules#pr-title-guide), the title (which becomes the commit message) determines the version bump for the packages you changed.

2. Wrap the text describing your change in more detail in the "CHANGELOG-DESCRIPTION" comment tags above, this is what will show up in the changelog!

3. DO NOT MERGE MANUALLY! When you are ready to merge and publish your changes, add the "Ship It" label to your Pull Request. This will trigger the merge process as long as all checks have completed, if the checks haven't completed the branch will be merged when they all pass.

**IMPORTANT:** If your PR contains breaking changes, please remember to follow the instructions for breaking changes!
-->

codecademydev commented:
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.3.52-alpha.023141.0<br/>@codecademy/gamut-labs@14.4.4-alpha.023141.0<br/>@codecademy/styleguide@36.4.4-alpha.023141.0<br/></pre><!--LERNA_VERSION_COMMENT-->

codecademydev commented:
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://608816dd8ba134699e4ee53f--gamut-preview.netlify.app>https://608816dd8ba134699e4ee53f--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/608816dd8ba134699e4ee53f>Deploy Logs</a></p><!--ALPHA_DEPLOY_COMMENT-->

JoshuaKGoldberg commented on packages/gamut-labs/src/ListSection/index.tsx:
```diff
@@ -0,0 +1,51 @@
+   */
+  renderItem: (item: T, index?: number) => ReactElement;
+  /**
+   * Number of items to be initally displayed (note: if the initial display amount matches the number of list items, Show All button will not appear).
```
Nit: [TSDoc](https://tsdoc.org/) has a think for this:

```suggestion
   * Number of items to be initally displayed
   *
   * @remarks
   * If the initial display amount matches the number of list items, Show All button will not appear.
```

JoshuaKGoldberg commented on packages/gamut-labs/src/ListSection/index.tsx:
```diff
@@ -0,0 +1,51 @@
+  /**
+   * Render function that will render each element in the list items.
+   */
+  renderItem: (item: T, index?: number) => ReactElement;
```
The `?` means this component _might_ provide an `index: number` when `renderItem` is called. I see in the code below that it's never provided, so you probably want this:

```suggestion
  renderItem: (item: T) => ReactElement;
```

JoshuaKGoldberg commented on packages/gamut-labs/src/ListSection/index.tsx:
```diff
@@ -0,0 +1,51 @@
+    return listItems.slice(0, displayLimit).map((item) => renderItem(item));
+  };
+
+  const renderfooterButton = () => {
```
Nit: capitalization?

```suggestion
  const renderFooterButton = () => {
```

This is part of why I personally prefer not to use inline `render*` functions; IMO they add a bit of extra overhead. Not a blocker from me, just a side comment.

sooinc replied:
Fixed the capitalization! Also agree that they are not always the best and can add extra overhead. I go back and forth but I think it works here

JoshuaKGoldberg commented on packages/styleguide/stories/Molecules/ListSection/index.stories.mdx:
```diff
@@ -0,0 +1,31 @@
+
+You will want to use this when the main content of your page is easily divided/organized into multiple sections. You would use one ListSection component to display a list of items in a section. It has a built in "Show All" functionality. "Show All" button will appear if the number of list items is higher than the initial display amount you wish to show. Once "Show All" button is clicked, it will  ...
+
+It is built on top of [Page Section](https://gamut.codecademy.com/storybook/?path=/docs/molecules-pagesection--page-section) so you may pass in optional header button and header secondary button.
```
Err, can you? The props aren't hooked up for that now.

sooinc replied:
sorry i missed this - i passed the optional props in now :)

JoshuaKGoldberg commented on packages/gamut-labs/src/ListSection/index.tsx:
```diff
@@ -0,0 +1,51 @@
+  initialDisplayAmount?: number;
+};
+
+export const ListSection = <T extends unknown>({
```
Is there a reason why this needs to use a `listItems` with its own `renderItem` and `T` generic to render children? If the point is to just make sure things are provided as an array, you could make the `children` prop typed as an array:

```tsx
children: React.ReactNode[];
```

That way consumers would have to provide an array as children, like:

```tsx
<ListSection title="Latest Certificate">
  {titles.map((title) => <Text>{title}</Text>}
</ListSection>
```

https://reactjs.org/docs/react-api.html#reactchildrentoarray is a thing too.

sooinc replied:
I went this route because I thought passing in a raw list of items and composing only the ones needed to show as react nodes was potentially better than composing all react nodes at once. 
But I don't actually know if we end up saving anything by doing that. I'm not opposed to using `children` prop!

ZackMFleischman replied:
@JoshuaKGoldberg The big reason to prefer passing in the items rather than an array of ReactNodes is because the `ListSection` component is in charge of how many of them to render (i.e. it renders only 3 by default, and then all of them if the user clicks "show all").

TIL about `Children.ToArray`. It looks like we could accomplish this via 
```
   renderListItems  = () => {
      if (showAll) return props.children;
      else return React.Children.toArray(children).slice(0, 3);
   }
```
I'm assuming the above will only kick off the useEffects for the first 3 items in the array since the rest aren't actually mounted? If so then this works well.

It would, as you said, rid us of the need for generic list item props and a renderer function.

JoshuaKGoldberg replied:
Yip yip, this looks good to me!

JoshuaKGoldberg replied:
Yip yip, regardless of what props this node receives as `children` or any other name, if it doesn't return them from the render they won't trigger effectcs.

JoshuaKGoldberg reviewed (changes requested):
Left a few small comments, with the one real blocker IMO being the usage of generic `renderItem` vs. `children`.

From the design perspective, we already have an accordion component set up with animations and everything. Can we start a conversation about expanding that for this kind of vertically flipped scenario?

codecademydev commented:
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.3.56-alpha.c61d97.0<br/>@codecademy/gamut-labs@14.5.2-alpha.c61d97.0<br/>@codecademy/styleguide@36.5.4-alpha.c61d97.0<br/></pre><!--LERNA_VERSION_COMMENT-->

codecademydev commented:
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60889a0e28c3d700c1d7e100--gamut-preview.netlify.app>https://60889a0e28c3d700c1d7e100--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60889a0e28c3d700c1d7e100>Deploy Logs</a></p><!--ALPHA_DEPLOY_COMMENT-->

sooinc commented:
@JoshuaKGoldberg re-requesting review as I addressed your comments.

RE: animation, connected with Timea and Cass and confirmed with Timea that for rendering additional items via "show all" animation is not needed (at least for now).

ZackMFleischman commented on packages/gamut-labs/src/ListSection/index.tsx:
```diff
@@ -0,0 +1,59 @@
+
+  const renderItems = () => {
+    const displayLimit = showAll ? listItems.length : initialDisplayAmount;
+    return listItems.slice(0, displayLimit);
```
nit: if there are a lot of items, this is copying the entire array unnecessarily.

ZackMFleischman commented on packages/gamut-labs/src/ListSection/__tests__/ListSection-test.tsx:
```diff
@@ -0,0 +1,66 @@
+
+const renderListSection = setupRtl(ListSection, listSectionProps(listItems));
+
+describe('ListSection', () => {
```
Does it make sense for us to have a default empty state if it has no items in the children array?

JoshuaKGoldberg replied:
This sounds perfect! 💯

ZackMFleischman reviewed (approved):
This looks good to me! Let's get webplatform approval before merging though. 

I did have the one question around an empty state though.

JoshuaKGoldberg commented on packages/gamut-labs/src/ListSection/index.tsx:
```diff
@@ -0,0 +1,59 @@
+
+export type ListSectionProps = {
+  title: string;
+  /**
+   * Children should be an array of react nodes.
+   */
```
This is self-explanatory from the type declaration ✨ 

```suggestion
```

JoshuaKGoldberg commented on packages/styleguide/stories/Molecules/ListSection.stories.mdx:
```diff
@@ -0,0 +1,55 @@
+
+You will want to use this when the main content of your page is easily divided/organized into multiple sections. You would use one ListSection component to display a list of items in a section. It has a built in "Show All" functionality. "Show All" button will appear if the number of list items is higher than the initial display amount you wish to show. Once "Show All" button is clicked, it will  ...
+
+It is built on top of [Page Section](https://gamut.codecademy.com/storybook/?path=/docs/molecules-pagesection--page-section) so you may pass in optional header button and header secondary button.
```
Maybe move this below the canvas and show this as a separate story to be extra clear? Up to you.

JoshuaKGoldberg reviewed (approved):
Thanks for being Gamut-first about this stuff! 🌮

dreamwasp reviewed (approved):
beautiful✨

codecademydev commented:
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.3.56-alpha.eeaece.0<br/>@codecademy/gamut-labs@14.5.2-alpha.eeaece.0<br/>@codecademy/styleguide@36.5.4-alpha.eeaece.0<br/></pre><!--LERNA_VERSION_COMMENT-->

codecademydev commented:
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6089884ae1ca262e87e0935b--gamut-preview.netlify.app>https://6089884ae1ca262e87e0935b--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6089884ae1ca262e87e0935b>Deploy Logs</a></p><!--ALPHA_DEPLOY_COMMENT-->

codecademydev commented:
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.3.56-alpha.ea6757.0<br/>@codecademy/gamut-labs@14.5.2-alpha.ea6757.0<br/>@codecademy/styleguide@36.5.4-alpha.ea6757.0<br/></pre><!--LERNA_VERSION_COMMENT-->

codecademydev commented:
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6089a1e09d12f40c9db2b120--gamut-preview.netlify.app>https://6089a1e09d12f40c9db2b120--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6089a1e09d12f40c9db2b120>Deploy Logs</a></p><!--ALPHA_DEPLOY_COMMENT-->

codecademydev commented:
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.3.57-alpha.3940a4.0<br/>@codecademy/gamut-labs@15.0.1-alpha.3940a4.0<br/>@codecademy/styleguide@36.5.5-alpha.3940a4.0<br/></pre><!--LERNA_VERSION_COMMENT-->

codecademydev commented:
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6089b43b9d12f4283db2b1c0--gamut-preview.netlify.app>https://6089b43b9d12f4283db2b1c0--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6089b43b9d12f4283db2b1c0>Deploy Logs</a></p><!--ALPHA_DEPLOY_COMMENT-->
